// @roamin/ecvrf/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateKey = exports.verify = exports.proofToHash = exports.prove = exports.keygen = exports.nonceGeneration = void 0;
const bn_js_1 = __importDefault(require("bn.js"));
const js_sha256_1 = require("js-sha256");
const elliptic = __importStar(require("elliptic"));
const utils = __importStar(require("minimalistic-crypto-utils"));
const EC = new elliptic.ec('secp256k1');
const suite = [0xfe];
function stringToPoint(s) {
    try {
        return EC.curve.decodePoint(s);
    }
    catch {
        return 'INVALID';
    }
}
function arbitraryStringToPoint(s) {
    if (s.length !== 32) {
        throw new Error('s should be 32 byte');
    }
    return stringToPoint(new Uint8Array([2, ...s]));
}
function hashToCurve(publicKey, alpha) {
    let hash = 'INVALID';
    let ctr = 0;
    while ((hash == 'INVALID' || hash.isInfinity()) && ctr < 256) {
        const hash_string = js_sha256_1.sha256
            .create()
            .update(suite)
            .update([0x01])
            .update(publicKey.encode('array', true))
            .update(alpha)
            .update([ctr])
            .digest();
        // @ts-ignore hash_string is a number[]
        hash = arbitraryStringToPoint(hash_string); // cofactor = 1, skip multiply
        ctr += 1;
    }
    if (hash == 'INVALID') {
        throw new Error('hashToCurve failed');
    }
    return hash;
}
function nonceGeneration(secretKey, h_string) {
    const h1 = js_sha256_1.sha256.array(h_string);
    let K = new Array(32).fill(0);
    let V = new Array(32).fill(1);
    K = js_sha256_1.sha256.hmac
        // @ts-ignore
        .create(K)
        .update(V)
        .update([0x00])
        .update(secretKey.toArray())
        .update(h1)
        .digest();
    V = js_sha256_1.sha256.hmac
        // @ts-ignore
        .create(K)
        .update(new Uint8Array(32).fill(1))
        .digest();
    K = js_sha256_1.sha256.hmac
        // @ts-ignore
        .create(K)
        .update(V)
        .update([0x01])
        .update(secretKey.toArray())
        .update(h1)
        .digest();
    V = js_sha256_1.sha256.hmac
        // @ts-ignore
        .create(K)
        .update(V)
        .digest();
    V = js_sha256_1.sha256.hmac
        // @ts-ignore
        .create(K)
        .update(V)
        .digest(); // qLen = hLen = 32, skip loop
    return new bn_js_1.default(V, 'hex');
}
exports.nonceGeneration = nonceGeneration;
function hashPoints(...points) {
    const str = [...suite, 0x02];
    for (const point of points) {
        str.push(...point.encode('array', true));
    }
    const c_string = js_sha256_1.sha256.digest(str);
    const truncated_c_string = c_string.slice(0, 16);
    const c = new bn_js_1.default(truncated_c_string);
    return c;
}
function decodeProof(pi) {
    const gamma_string = pi.slice(0, 33);
    const c_string = pi.slice(33, 33 + 16);
    const s_string = pi.slice(33 + 16, 33 + 16 + 32);
    const Gamma = stringToPoint(gamma_string);
    if (Gamma == 'INVALID') {
        return 'INVALID';
    }
    const c = new bn_js_1.default(c_string);
    const s = new bn_js_1.default(s_string);
    return {
        Gamma,
        c,
        s,
    };
}
function _prove(secretKey, alpha) {
    const publicKey = EC.keyFromPrivate(secretKey.toArray()).getPublic();
    const H = hashToCurve(publicKey, alpha);
    const h_string = H.encode('array', true);
    const Gamma = H.mul(secretKey);
    // @ts-ignore h_string is a number[]
    const k = nonceGeneration(secretKey, h_string);
    const c = hashPoints(H, Gamma, EC.g.mul(k), H.mul(k));
    const s = k.add(c.mul(secretKey)).umod(EC.n);
    const pi = [
        ...Gamma.encode('array', true),
        ...c.toArray('be', 16),
        ...s.toArray('be', 32),
    ];
    return pi;
}
function _proofToHash(pi) {
    const D = decodeProof(pi);
    if (D == 'INVALID') {
        throw new Error('Invalid proof');
    }
    const { Gamma } = D;
    const beta = js_sha256_1.sha256
        .create()
        .update(suite)
        .update([0x03])
        .update(Gamma.encode('array', true))
        .digest();
    return beta;
}
function _verify(publicKey, pi, alpha) {
    const D = decodeProof(pi);
    if (D == 'INVALID') {
        throw new Error('Invalid proof');
    }
    const { Gamma, c, s } = D;
    const H = hashToCurve(publicKey, alpha);
    const U = EC.g.mul(s).add(publicKey.mul(c).neg());
    const V = H.mul(s).add(Gamma.mul(c).neg());
    const c2 = hashPoints(H, Gamma, U, V);
    if (!c.eq(c2)) {
        throw new Error('Invalid proof');
    }
    return _proofToHash(pi);
}
function _validateKey(publicKey_string) {
    const publicKey = stringToPoint(publicKey_string);
    if (publicKey == 'INVALID' || publicKey.isInfinity()) {
        throw new Error('Invalid public key');
    }
    return publicKey;
}
function keygen() {
    const keypair = EC.genKeyPair();
    const secretKey = keypair.getPrivate('hex');
    const publicKey = keypair.getPublic('hex');
    return {
        secretKey,
        publicKey,
    };
}
exports.keygen = keygen;
/**
  * Generates proof from a secret key and message
  * @param secretKey the secret key to use to generate the proof (hex string)
  * @param alpha the message to use to generate the proof (hex string)
  * @returns the proof as an hex string
  * @example
  * ```js
  * const { prove, proofToHash, verify } = require('@roamin/ecvrf');
  * const elliptic = require('elliptic');
  *
  * const EC = new elliptic.ec('secp256k1');
  *
  * const SECRET = EC.keyPair({ priv: 'c9afa9d845ba75166b5c215767b1d6934e50c3db36e89b127b8a622b120f6721', privEnc: 'hex' });
  *
  * const msg = Buffer.from('sample').toString('hex');
  *
  * // VRF proof and hash output
  * const proof = prove(SECRET.getPrivate(), msg);
  * ```
  */
function prove(secretKey, alpha) {
    const pi = _prove(new bn_js_1.default(secretKey, 'hex'), utils.toArray(alpha, 'hex'));
    return utils.toHex(pi);
}
exports.prove = prove;
/**
  * Generates the hash of a proof
  * @param pi the proof to hash (hex string)
  * @returns the hash proof as an hex string
  * @example
  * ```js
  * const { prove, proofToHash, verify } = require('@roamin/ecvrf');
  * const elliptic = require('elliptic');
  *
  * const EC = new elliptic.ec('secp256k1');
  *
  * const SECRET = EC.keyPair({ priv: 'c9afa9d845ba75166b5c215767b1d6934e50c3db36e89b127b8a622b120f6721', privEnc: 'hex' });
  *
  * const msg = Buffer.from('sample').toString('hex');
  *
  * // VRF proof and hash output
  * const proof = prove(SECRET.getPrivate(), msg);
  * const hash = proofToHash(proof);
  * ```
  */
function proofToHash(pi) {
    const beta = _proofToHash(utils.toArray(pi, 'hex'));
    return utils.toHex(beta);
}
exports.proofToHash = proofToHash;
/**
  * Verifies the provided VRF proof and computes the VRF hash output
  * @param publicKey the public key to use to verify the proof (hex string)
  * @param pi the proof to verify (hex string)
  * @param alpha the message to verify (hex string)
  * @returns the hash proof as an hex string
  * @example
  * ```js
  * const { prove, proofToHash, verify } = require('@roamin/ecvrf');
  * const elliptic = require('elliptic');
  *
  * const EC = new elliptic.ec('secp256k1');
  *
  * const SECRET = EC.keyPair({ priv: 'c9afa9d845ba75166b5c215767b1d6934e50c3db36e89b127b8a622b120f6721', privEnc: 'hex' });
  *
  * const msg = Buffer.from('sample').toString('hex');
  *
  * // VRF proof and hash output
  * const proof = prove(SECRET.getPrivate(), msg);
  * const hash = proofToHash(proof);
  *
  * // VRF proof verification (returns VRF hash output)
  * const beta = verify(SECRET.getPublic('hex'), proof, msg);
  * ```
  */
function verify(publicKey, pi, alpha) {
    const beta = _verify(EC.curve.decodePoint(publicKey, 'hex'), utils.toArray(pi, 'hex'), utils.toArray(alpha, 'hex'));
    return utils.toHex(beta);
}
exports.verify = verify;
function validateKey(publicKey) {
    _validateKey(utils.toArray(publicKey, 'hex'));
    return;
}
exports.validateKey = validateKey;
